src/transformers/models/cogvlm/modeling_cogvlm.py

```diff
@@ -0,0 +1,1053 @@
+        >>> import torch
+        >>> import requests
+        >>> from PIL import Image
+
+        >>> processor = CogVLMProcessor.from_pretrained("THUDM/cogvlm-chat-hf")
+        >>> model = CogVLMModel.from_pretrained("THUDM/cogvlm-chat-hf")
+
+        >>> # load image
+        >>> url = "http://images.cocodataset.org/val2017/000000039769.jpg"
+        >>> image = Image.open(requests.get(url, stream=True).raw)
+        >>> query = "Describe this image"
+
+        >>> prompt = f"Question: {query} Answer:"
+        >>> inputs = processor(images=image, text=prompt, return_tensors="pt")
+
+        >>> # forward pass
+        >>> with torch.no_grad():
+        ...     outputs = model(**inputs)
+
+        >>> last_hidden_state = outputs.last_hidden_state
+        ```
+        """
+        output_attentions = output_attentions if output_attentions is not None else self.config.output_attentions
+        output_hidden_states = (
+            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
+        )
+        use_cache = use_cache if use_cache is not None else self.config.use_cache
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        if past_key_values is not None:
+            # generate mode with past_key_values. the image features are already mapped
+            # update attention_mask
+            if pixel_values is not None:
+                batch_size = pixel_values.shape[0]
+                vision_mask = torch.full(
+                    size=(batch_size, self.num_vision_tokens), fill_value=1, device=attention_mask.device
+                )
+                attention_mask = torch.cat(
+                    [attention_mask[:, :-1], vision_mask, attention_mask[:, -1].repeat(batch_size, 1)], dim=1
+                )
+        else:
+            # not allow for inputs_embeds, because we want to process image feature
+            # assert input_ids is not None and inputs_embeds is None, f"{input_ids} {inputs_embeds}"
+
+            if pixel_values is not None and input_ids is not None:
+                # multi-modality
+                if len(input_ids) != len(pixel_values):
+                    raise ValueError("Make sure to pass as many texts as images")
+
+                # prepend the input_ids and token_type_ids with image tokens
+                batch_size = input_ids.shape[0]
+
+                vision_input_ids = (
+                    torch.tensor([self.config.bos_token_id] + [self.config.pad_token_id] * self.num_vision_tokens)
+                    .repeat(batch_size, 1)
+                    .to(input_ids.device)
+                )
+                vision_token_type_ids = (
+                    torch.tensor([LANGUAGE_TOKEN_TYPE] + [VISION_TOKEN_TYPE] * self.num_vision_tokens)
+                    .repeat(batch_size, 1)
+                    .to(token_type_ids.device)
+                )
```

Comment:
split in a few lines

Reply:
these tensors will be created for each foward, not super super optimal. You can store a buffer for that no? and just expand it

Reply:
Yes, although they aren't recreated when using a cache (which is used by default)